### PR TITLE
[qat]A more stable conv_bn fusion for qat training.

### PR DESCRIPTION
**Pull request by newstzpz**
Summary:
A more stable conv_bn fusion for qat training:
* Existing implementation may cause QAT training loss become NaN. This could happen when the fused conv for qat (torch/nn/intrinsic/qat/modules/conv_fused.py) is used and is independent of if fake_quant is enabled.
  * This is caused by the unscaling for the conv output (`conv_orig = conv / scale_factor` where `scale_factor = bn.weight / running_std`) when there is 0 in `bn.weight`.

* This implementation follows the [white paper](https://arxiv.org/pdf/1806.08342.pdf) better and fixed the issue by scaling `running_std / std_Y` instead and compute the fused output accordingly (see comments in conv_fused.py for more details):
  * It comes at the cost of running conv twice (one to update bn statistics and one to compute fake quant for fused weights).
  * It does not need to use conv bias for back prop.
  * It uses the bn statistics computed with the current input batch, while the existing code uses the statistics without the current batch.
* The implementation could be enabled by setting the flag `_enable_slow_path_for_better_numerical_stability` to True after the model is prepared for QAT.

* Unit test
  * Added test case for zero `bn.weight`.
  * Added test case for conv to has bias.

Test Plan: buck run mode/dev-nosan //caffe2/test:quantization -- -r quantization.eager.test_quantize_eager_qat

Differential Revision: D29506778

### CODE REVIEW

File: test/quantization/eager/test_quantize_eager_qat.py [hunk 1]
```diff
@@ -749,7 +749,10 @@ def forward(self, x):
            use_relu=st.booleans(),
            eps=st.sampled_from([1e-5, 1e-4, 1e-3]),
            momentum=st.sampled_from([0.1, 0.2, 0.3]),
-           freeze_bn=st.booleans())
+           freeze_bn=st.booleans(),
+           zero_gamma=st.booleans(),
+           has_bias=st.booleans(),
+           use_slow_fusion=st.booleans())
```
Comment (vkuzo) [inline test/quantization/eager/test_quantize_eager_qat.py:755]: just checking, have we tested the new path with real inputs?  We don't really trust hypothesis anymore to give us realistic inputs and we plan to deprecate it.  I'm assuming you have since you mentioned that various people already use this improvement in their models, but just checking.

**Review by vkuzo — APPROVED** [review]

File: test/quantization/eager/test_quantize_eager_qat.py (hunk 1 above)
Comment (newstzpz) [reply test/quantization/eager/test_quantize_eager_qat.py:755]: Yes, we have tested it before. Also if you remove line 817, the test will fail as well.
